### PR TITLE
Support `RustString` in extern Swift functions

### PR DESCRIPTION
**Pull request by NiwakaDev**
Related to #201. This commit adds support `RustString`  in extern Swift functions.

Here's an example of using this.

```Swift
//Swift
func reflect_swift_string(arg: RustString) -> RustString {
    arg
}

```

```Rust
//Rust
#[swift_bridge::bridge]
mod ffi {
    extern "Swift" {
        fn reflect_swift_string(arg: String) -> String;
    }
}

let foo = "foo";
let string = ffi::reflect_swift_string(foo.to_string());
assert_eq!(string.len(), 3);
assert_eq!(&string, foo);
```

### CODE REVIEW

File: crates/swift-integration-tests/src/string.rs
Comment (NiwakaDev) [inline crates/swift-integration-tests/src/string.rs:16]: Maybe, if this PR get merged, we could remove this from `crates/swift-integration-tests/src/string.rs`.

File: crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs
Comment (chinedufn) [inline crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs:294]: Why do we need to call `intoRustString` if, according to your PR body, `some_function` would return a `RustString` here?

File: crates/swift-integration-tests/src/string.rs
Comment (chinedufn) [reply crates/swift-integration-tests/src/string.rs:16]: Makes sense

**Review by chinedufn — COMMENTED** [review]
Left one question.

Tests are failing, but it seems like it started here https://github.com/chinedufn/swift-bridge/commit/7c14ea07ab71702019e652fcc5a5f7a111d5ea0e

The PR was passing when I merged it so maybe it conflicted with something in the master branch..

**Comment by chinedufn** [comment]
Mind adding an example bridge module to your PR body so that it's more clear that the bridge module uses `String` and not `RustString`

**Comment by NiwakaDev** [comment]
>Left one question.
>Tests are failing, but it seems like it started here https://github.com/chinedufn/swift-bridge/commit/7c14ea07ab71702019e652fcc5a5f7a111d5ea0e
>The PR was passing when I merged it so maybe it conflicted with something in the master branch..

By the way, Tests also failed before #218 got merged.

**Comment by chinedufn** [comment]
Hmmm gotcha. I can't dive in for a couple days. Any idea where the issue might be stemming from?

**Comment by NiwakaDev** [comment]
>Any idea where the issue might be stemming from?

I don't understand what causes the failure. I'll investigate it this week.

It might be a good idea to start by addressing the compiler warnings, like so:
```
/Users/runner/work/swift-bridge/swift-bridge/SwiftRustIntegrationTestRunner/Generated/swift-integration-tests/swift-integration-tests.h:741:232: warning: declaration of 'struct __swift_bridge__$ResultAsyncResultOkEnumAndAsyncResultOpaqueRustType1' will not be visible outside of this function
```

**Comment by NiwakaDev** [comment]
I found that what cause the failure. I'll make a PR for fixing this problem tomorrow.

**Comment by NiwakaDev** [comment]
I found that #213 causes the failure.

## Detail

Let's say that we have a tuple type and a result type, like so:

```rust
mod ffi {
    extern "Rust" {
        type ResultTestOpaqueRustType;
    }

    extern "Rust" {
        fn rust_func_return_result_tuple_transparent_enum(
            succeed: bool,
        ) -> Result<(i32, ResultTestOpaqueRustType, String), i32>;
    }
}
```

Will potentiality generate:

```C
struct __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32 __swift_bridge__$rust_func_return_result_tuple_transparent_enum(bool succeed);
typedef enum __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Tag {__swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$ResultOk, __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$ResultErr} __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Tag;
union __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Fields {struct __swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString ok; int32_t err;};
typedef struct __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32{__swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Tag tag; union __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Fields payload;} __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32;
typedef struct __swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString { int32_t _0; void* _1; void* _2; } __swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString;           
```

Or 

```C
struct __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32 __swift_bridge__$rust_func_return_result_tuple_transparent_enum(bool succeed);
typedef struct __swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString { int32_t _0; void* _1; void* _2; } __swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString;
typedef enum __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Tag {__swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$ResultOk, __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$ResultErr} __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Tag;
union __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Fields {struct __swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString ok; int32_t err;};
typedef struct __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32{__swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Tag tag; union __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Fields payload;} __swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32;
```

The former causes the failure because `__swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Fields` couldn't know `__swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString`, on the other hand the later doesn't cause the failure.

The reason for the non-deterministic ordering is due to `HashSet<String>`:

https://github.com/chinedufn/swift-bridge/blob/7c14ea07ab71702019e652fcc5a5f7a111d5ea0e/crates/swift-bridge-ir/src/codegen/generate_c_header.rs#L367-L370

And `Vec<TokenStream>` I introduced in #213:

https://github.com/chinedufn/swift-bridge/blob/7c14ea07ab71702019e652fcc5a5f7a111d5ea0e/crates/swift-bridge-ir/src/bridged_type/bridgeable_result.rs#L384

## Potential Solution

### 1
`__swift_bridge__$ResultTupleI32ResultTestOpaqueRustTypeStringAndI32$Fields` needs to know `__swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString`, but `__swift_bridge__$tuple$I32ResultTestOpaqueRustTypeString` could be generated without any dependencies. This means that we need to think the order of generating the two declarations. Maybe by introducing  something like a directed graph or that to represent the type dependencies, we could fix this problem.

### 2

Since Solution 1 might be complicated, so maybe we could stop support #213 to fix the failure.

What do you think about this?

**Comment by chinedufn** [comment]
Thanks for the detailed breakdown and links. This made it very easy for me to understand this complex issue.

I don't think we need a complex solution.

I think we can solve this with the approach described here https://github.com/chinedufn/swift-bridge/issues/170#issuecomment-1431588845

---

In essence, we use a `parse_queue: Vec<&TypeDeclaration>` and a `encountered: HashSet<String>`

We iterate over the types and push them to the parse queue. If the type contains other types we push those other types to the parse queue first, recursively.

Each time we push a type we insert its C header code into the `encountered` `HashSet`, this way we don't push the same type twice.

So, this guarantees that a type always appears after all of the types that we depend on.

---

So, because the solution should be fairly self-contained and not require much code, I like potential solution 1 the best where we just fix the problem.

---

Does that make sense?

File: crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs
Comment (chinedufn) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs:294]: @NiwakaDev waiting on this question then after that we can merge this and solve the test stuff in a separate PR

Comment (chinedufn) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs:294]: Oh and also please add an example bridge module to your PR body so that it's more clear that the bridge module uses String and not RustString

**Comment by NiwakaDev** [comment]
Thank you for your quick reply.

>Each time we push a type we insert its C header code into the encountered HashSet, this way we don't push the same type twice.
So, this guarantees that a type always appears after all of the types that we depend on.

I guess that the contents order of `HashSet` isn't deterministic. Why does this guarantees it?.

**Comment by NiwakaDev** [comment]
>Does that make sense?

Yes. Thanks!!

File: crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs
Comment (chinedufn) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs:294]: Any availability to look at this? One this is addressed I can merge

Comment (NiwakaDev) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs:294]: I am working on it, but I'm slightly busy this month. Sorry...

File: SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunner/String.swift
Comment (NiwakaDev) [inline SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunner/String.swift:16]: @chinedufn 

`create_swift_string` returns `String`. To maintain consistency and remove `toRustString`(`extern_swift_func_takes_and_returns_string`), I changed `take_and_return_swift_string` return value from `RustString` to `String`. But it is inefficient. what do you think about this?

Comment (NiwakaDev) [reply SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunner/String.swift:16]: Here're `create_swift_string` and `take_and_return_swift_string` implementation.
```String
//
//  String.swift
//  SwiftRustIntegrationTestRunner
//
//  Created by Frankie Nwafili on 2/18/22.
//

import Foundation

func create_swift_string() -> String {
    "hello"
}

func take_and_return_swift_string(arg: RustString) -> String {
    arg.toString()
}

```

Comment (chinedufn) [reply SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunner/String.swift:16]: I haven't reviewed the code yet but this sounds fine.

Let's just document this behavior in the `String` chapter in the book so that users are aware of this implicit clone

File: crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs
Comment (chinedufn) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/string_codegen_tests.rs:294]: > Why do we need to call intoRustString if, according to your PR body, some_function would return a RustString here?

Nevermind, the reason is that we support returning anything that can be converted into a RustString.

File: crates/swift-integration-tests/src/string.rs
Comment (chinedufn) [inline crates/swift-integration-tests/src/string.rs:11]: Let's call this `take_and_return_rust_string` and make the Swift function `(arg: RustString) -> RustString`

This way we have a test for both `-> String` and `-> RustString`

**Review by chinedufn — APPROVED** [review]
Left minor feedback then looks good to me.

Can you

1. Address feedback
2. Update your PR body to show an example bridge module

After that you can merge this PR.

Thanks!

**Comment by NiwakaDev** [comment]
@chinedufn 

Addressed your review. Please let me know if there are any errors or omissions.

**Comment by chinedufn** [comment]
> Please let me know if there are any errors or omissions.

Minor, but the function should be called `reflect_rust_string` (not `reflect_swift_string`).

After that LGTM

**Comment by NiwakaDev** [comment]
@chinedufn 

Done.

**Comment by chinedufn** [comment]
Thanks!

**Pull request merged**
